packages/@react-spectrum/actionbar/src/ActionBar.tsx

```diff
@@ -82,6 +82,53 @@ const ActionBarInner = React.forwardRef((props: ActionBarInnerProps, ref: DOMRef
     announce(formatMessage('actionsAvailable'));
   }, [formatMessage]);
 
+  let restoreFocusRef = useRef<HTMLElement>(null);
+  let containerRef = useRef<HTMLElement>(null);
+
+  useEffect(() => {
+    // Event handler to keep track of focus changes within ActionBarContainer when ActionBar is open.
+    let onFocusContainer = (e:FocusEvent) => {
+      let activeElement = e.target as HTMLElement;
+      if (domRef.current && !domRef.current.contains(activeElement)) {
+        restoreFocusRef.current = activeElement;
+      }
+    };
```

Comment:
This is very close. One last thing I noticed was if the user tabs into the ActionBar and then scrolls the focused key out of view in such a way that it is removed from the DOM, closing the ActionBar doesn't restore focus to the previously focused cell. This is because the restoreFocusRef (i.e. the focused cell) is no longer in the DOM and onFocusContainer doesn't update the restoreFocusRef because the active element is one of the buttons in the ActionBar when the focused key is scrolled out of view.

Perhaps we should bring back the fallback treewalker code when `document.body.contains(restoreFocusRef.current)` returns false in the useEffect cleanup?
```
        let walker = getFocusableTreeWalker(container, {tabbable: true});
        let node = walker.nextNode() as HTMLElement;
        node && node.focus();
```

Repro steps:
1. Select the first row of the ActionBar's TableView via arrow keys + Enter
2. Tab into the ActionBar
3. Scroll to the last row of the TableView via touchpad scroll/scroll wheel
4. Close the ActionBar via escape key or focusing the 'x' and hitting enter. Note focus isn't restored

Reply:
Unsure if it's the same reason, but you can also use the default story
1. navigate to the 10th item, select it
2. navigate to the 5th item, select it
3. tab to the actionbar
4. press Esc

for me, focus went to item 12, though I think it should have gone to either 5? given the discussion?

Reply:
Hm I'm having a hard time reproducing in https://reactspectrum.blob.core.windows.net/reactspectrum/33780525ea16cfe5c1c0a87dfba2b8cf855901b0/storybook/index.html?path=/story/actionbar--default. Can you still reproduce?

Reply:
Ah nvm, I'm seeing it in the react 17 storybook, hmmm

Reply:
Looks like the issue is when the user tabs out of the TableView and into the ActionBar, `useSelectableCollection` will [focus the last tabbable item](https://github.com/adobe/react-spectrum/blob/b34e500ddd6ca860cbc943f90a85ea911f2d881b/packages/%40react-aria/selection/src/useSelectableCollection.ts#L237-L266) so that we can exit the table as a single tab stop. This gets picked up by `onFocusContainer` and is set as the element to restore focus to hence the behavior observed by @snowystinger.

Reply:
@majornista Ah scratch that, perhaps the fix is just to remove the `restoreFocus` prop from ActionBar's FocusScope. My previous comment is a red herring I think since `useSelectableItem` will properly move the focus back to the tracked focusedKey I believe via its useEffect. However, FocusScope's `restoreFocus` is problematic here because it tries to restore focus back to the item that was focused when the ActionBar first mounted but that isn't necessary the right place to restore focus to since the user can move focus before tabbing into the ActionBar. Since we are basically handling restoring focus via the ActionBar useEffect logic we can do away with the ActionBar's FocusScope `restoreFocus` IMO.